src/box/sql/mem.c

```diff
@@ -2034,15 +2039,86 @@ mem_add(const struct Mem *left, const struct Mem *right, struct Mem *result)
 	return 0;
 }
 
+/**
+ * Add the first MEM to the second MEM and write the result to the third MEM.
+ * The first MEM should be of DATETIME type and the second MEMs should be of
+ * INTERVAL type. The result is of DATETIME type.
+ */
+static int
+mem_add_dt(const struct Mem *left, const struct Mem *right, struct Mem *result)
+{
+	assert(mem_is_datetime(left) && !mem_is_metatype(left));
+	if (!mem_is_interval(right) || mem_is_metatype(right)) {
+		diag_set(ClientError, ER_SQL_TYPE_MISMATCH, mem_str(right),
+			 "interval");
+		return -1;
+	}
+	struct datetime res = left->u.dt;
+	if (datetime_increment_by(&res, 1, &right->u.itv) != 0)
+		return -1;
+	mem_set_datetime(result, &res);
+	return 0;
+}
+
+/**
+ * Add the first MEM to the second MEM and write the result to the third MEM.
+ * The first MEM should be of INTERVAL type and the second MEMs should be of
+ * INTERVAL or DATETIME type. The result is of the same type as the second
+ * argument.
+ */
+static int
+mem_add_itv(const struct Mem *left, const struct Mem *right, struct Mem *result)
+{
+	assert(mem_is_interval(left) && !mem_is_metatype(left));
+	if (!(mem_is_interval(right) || mem_is_datetime(right)) ||
+	    mem_is_metatype(right)) {
+		diag_set(ClientError, ER_SQL_TYPE_MISMATCH, mem_str(right),
+			 "datetime or interval");
+		return -1;
+	}
+	if (mem_is_datetime(right))
+		return mem_add_dt(right, left, result);
+	struct interval res = left->u.itv;
+	if (interval_interval_add(&res, &right->u.itv) != 0)
+		return -1;
+	mem_set_interval(result, &res);
+	return 0;
+}
```

Comment:
You can simplify this all a bit (same for the subtraction):
```Diff
diff --git a/src/box/sql/mem.c b/src/box/sql/mem.c
index bc9f302cc..063ea69ab 100644
--- a/src/box/sql/mem.c
+++ b/src/box/sql/mem.c
@@ -2053,11 +2053,8 @@ mem_add_dt(const struct Mem *left, const struct Mem *right, struct Mem *result)
 			 "interval");
 		return -1;
 	}
-	struct datetime res = left->u.dt;
-	if (datetime_increment_by(&res, 1, &right->u.itv) != 0)
-		return -1;
-	mem_set_datetime(result, &res);
-	return 0;
+	mem_set_datetime(result, &left->u.dt);
+	return datetime_increment_by(&result->u.dt, 1, &right->u.itv);
 }
 
 /**
@@ -2070,19 +2067,15 @@ static int
 mem_add_itv(const struct Mem *left, const struct Mem *right, struct Mem *result)
 {
 	assert(mem_is_interval(left) && !mem_is_metatype(left));
-	if (!(mem_is_interval(right) || mem_is_datetime(right)) ||
-	    mem_is_metatype(right)) {
+	if (mem_is_datetime(right))
+		return mem_add_dt(right, left, result);
+	if (!mem_is_interval(right) || mem_is_metatype(right)) {
 		diag_set(ClientError, ER_SQL_TYPE_MISMATCH, mem_str(right),
 			 "datetime or interval");
 		return -1;
 	}
-	if (mem_is_datetime(right))
-		return mem_add_dt(right, left, result);
-	struct interval res = left->u.itv;
-	if (interval_interval_add(&res, &right->u.itv) != 0)
-		return -1;
-	mem_set_interval(result, &res);
-	return 0;
+	mem_set_interval(result, &left->u.itv);
+	return interval_interval_add(&result->u.itv, &right->u.itv);
 }
```

Reply:
Thank you for the diff. I applied, but I added one more condition in one of `if`s:
```
+	if (mem_is_datetime(right) && !(mem_is_metatype(right))
+		return mem_add_dt(right, left, result);
```

Reply:
In the diff I didn't change the subtractions. I only mentioned them in the comment. Lets do the same for them.